### ISSUE TITLE
Circular dependencies cause code duplication in bundle on Windows

### ISSUE THREAD

**Issue by Septh** [wessberg/rollup-plugin-ts#23]
_(full repro here: https://github.com/Septh/wessberg-ts-plugin-issue-repro)_

Consider the following project:

```ts
// index.ts
import * as math from './math'

export class Test {
    public static defaultFactor = 1
    public mul() {
        return math.mult(2)
    }
}

const t = new Test()
console.log(t.mul())
```
```ts
// math.ts
import { Test } from './index'    // <-- circular dep

export const double = (x: number) => x * 2
export const mult = (x: number, y: number = Test.defaultFactor) => x * y

```

When compiled and bundled, this produces the following code:
```js
var Test = /** @class */ (function () {
    function Test() {
    }
    Test.prototype.mul = function () {
        return mult(2);
    };
    Test.defaultFactor = 1;
    return Test;
}());
var t = new Test();
console.log(t.mul());

var mult = function (x, y) {
    if (y === void 0) { y = Test.defaultFactor; }
    return x * y;
};

var Test$1 = /** @class */ (function () {
    function Test() {
    }
    Test.prototype.mul = function () {
        return mult(2);
    };
    Test.defaultFactor = 1;
    return Test;
}());
var t$1 = new Test$1();
console.log(t$1.mul());

export { Test$1 as Test };
```

Class `Test` is emitted twice and what's worse, so is the instantiated var `t`.

I've checked that this is not an issue with Rollup: the same code written in plain JavaScript produces correct result (see `bundle-js.js` in repro). This doesn't seem to be an issue with TypeScript neither since running `tsc` with the `--outFile` option produces correct result too.

Also, FWIW, the official `rollup-plugin-typescript` has the same issue 😀

**Comment by wessberg** [comment]
Hi there!
The reason why it's not an issue in `bundle-js.js` is that you're looking at non-transpiled _ES2015esque_ syntax there. It's not an issue in the tsc bundle either since it's not actually flattened, it just contains AMD definitions inlined in the same file. However, the way the Typescript compiler emits to es5 (by assigning the result of a self-invoking function to a hoisted variable is what makes Rollup have a hard time figuring out of the symbol needs duplication or not. You can reproduce the behavior by running tsc without outFile to produce a new ESM bundle (es5 transpired) and run _that_ through Rollup to see the same thing happen. Rollup does this because it sometimes cannot reliably detect if a symbol/identifier is unique

**Comment by Septh** [comment]
Oh, my bad: I wanted to make this report simple and forgot about es5 ~~downgrading~~ transpiling :)

Actually, I target es2018 in my project and I'm not using any transpiling at all. I updated the repro to target `esnext` in all 3 configs and the results are the same so I don't think it is an issue with es5.

**Comment by Septh** [comment]
I tried your suggestion of going though `tsc --outdir` (instead of outfile) before bundling: since tsc generates _exactly_ the same code than the one I wrote by hand in the `test-js` folder, Rollup bundles it fine. So, in short:
```
source -> tsc -> rollup w/o plugin = bundle ok
source -> rollup w/plugin (yours or rollup-typescript-plugin) = bundle wrong
```

**Comment by Septh** [comment]
I think I found what's going wrong here.

I wrote a little plugin that hooks on `writeBundle` to print the final contents of the bundle to the console. It appears that `index.ts` is included twice:

![1](https://user-images.githubusercontent.com/11949623/60849778-97a46d00-a1ec-11e9-8c54-2e8c31e534de.png)

But as you can see on the screenshot, the path of the second instance is not normalized, so Rollup thinks these are two different files.

So I extended my little plugin to hook on `resolveId` as well to make sure all paths are properly normalized and... tada!

![2](https://user-images.githubusercontent.com/11949623/60849945-68423000-a1ed-11e9-80ba-b933abaa8cfc.png)

`index.ts` is now included only once, no more duplicated code in the bundle.

I updated [my repro](https://github.com/Septh/wessberg-ts-plugin-issue-repro) so you can see by yourself. I think this is something that [c|sh]ould be taken care of by rollup-plugin-ts.

**Comment by wessberg** [comment]
Thanks for looking into this! You are correct - this behavior is not present on UNIX systems where POSIX filenames are being used.

Here’s the output from Mac OS for example:

```javascript
const mult = (x, y = Test.defaultFactor) => x * y;

class Test {
    mul() {
        return mult(2);
    }
}
Test.defaultFactor = 1;
const t = new Test();
console.log(t.mul());

export { Test };
```

I’ve had some issues with Windows in the past because of the differences in the filesystem (see https://github.com/wessberg/rollup-plugin-ts/issues/21 and https://github.com/wessberg/rollup-plugin-ts/issues/19), but these issues have been resolved.

Thanks to your repro and your solution, I’ve fixed this issue by normalizing resolved paths before returning them to Rollup (because Rollup doesn’t seem to normalize paths returned from resolveId).

This has been released in v1.1.61
